src/ogma/features/nodeGrouping.ts

```diff
@@ -0,0 +1,205 @@
+        },
+        showContents: true,
+        duration: 300,
+        padding: 10
+      });
+      this._listenToTransformationEvents();
+    } else {
+      await this.refreshTransformation();
+    }
+  }
+
+  /**
+   * refresh the transformation
+   * Called when there is a change in the rule
+   */
+  public async refreshTransformation(): Promise<void> {
+    await this.transformation?.refresh();
+  }
+
+  /**
+   * init node grouping style
+   */
+  public initNodeGroupingStyle(): void {
+    this.nodeGroupingStyleRule = this._ogma.styles.addRule({
+      nodeAttributes: {
+        // Any default style will go here
+        text: {
+          content: (node: Node<LkNodeData> | undefined): string | undefined => {
+            return this._getNodeGroupingCaption(node);
+          },
+          style: 'bold'
+        },
+        layer: -1,
+        opacity: 0.32
+      },
+      nodeSelector: (node) => {
+        return node.isVirtual();
+      },
+      nodeDependencies: {self: {data: true}}
+    });
+  }
+
+  /**
+   * run layout on all subnodes of virtual nodes
+   */
+  public async runLayoutOnAllSubeNodes(): Promise<void> {
+    // @ts-ignore getContext exists on the transformation but hidden by the types
+    const virtualNodes = this.transformation.getContext().virtualNodes;
+    const rawNodesList = virtualNodes.getSubNodes();
+    const promisesList: Promise<void>[] = [];
+    for (let i = 0; i < rawNodesList.length; i++) {
+      const subNodes = rawNodesList[i];
+      if (subNodes !== undefined) {
+        promisesList.push(this._runSubNodesLayout(subNodes));
+      }
+    }
+    promisesList.push(this._runSubNodesLayout(virtualNodes));
+    await Promise.all(promisesList);
+  }
+
+  /**
+   * Run the layout on the subnodes of the virtual node
+   * @param subNodes nodes part of a virtual node
+   */
+  private async _runSubNodesLayout(subNodes: NodeList<LkNodeData, LkEdgeData>): Promise<void> {
+    if (subNodes.size === 0) {
+      return;
+    }
+
+    const noEdges = subNodes.getAdjacentEdges({bothExtremities: true}).size === 0;
+    if (noEdges) {
+      await this._runCirclePack(subNodes);
+    } else {
+      await this._runForceLayout(subNodes);
+    }
+  }
+
+  /**
+   * Return the caption of a virtual node
+   * @param node reference to the virtual node
+   */
+  private _getNodeGroupingCaption(node: Node<LkNodeData> | undefined): string | undefined {
+    if (node !== undefined && node.isVirtual()) {
+      const size = node.getSubNodes()!.filter((e) => !e.hasClass('filtered')).size;
+      return `${this.groupRule?.ruleName} - ${size}`;
+    }
+  }
+
+  /**
+   * Listen to transformationEnabled and transformationRefresh events to run the layout on the subnodes
+   */
+  private _listenToTransformationEvents(): void {
+    this._ogma.events.on(
+      ['transformationEnabled', 'transformationRefresh'],
+      async (transformations) => {
+        if (transformations.target.getId() === this.transformation?.getId()) {
```

Comment:
when is this false?

Reply:
if you have and refresh another transformation. For example edge grouping